blockchain/blockchain_data_structure.py

```diff
@@ -1,59 +1,101 @@
+
+        # What type should from and to address be? How do we define public keys?
+        # TODO - Add from and to address type checks
+
+        # What type we want amount to be? Probably not int
+        # TODO - Add amount type check
+
+        if amount < 0:
+            raise Exception("Transaction amount must be greater or equal than 0")
+
     def calculate_hash(self):
         t_hash = SHA256.new()
-        t_hash.update("self.id + self.fromAddress + self.toAddress + str(self.amount)".encode())
+        t_hash.update(self.transaction_content().encode())
         return t_hash
 
-    def sign_transaction(self, node_id):
-        self.signature = sign_hash(self.calculate_hash(), node_id)
-
-    def check_valid(self, node_id):
+    def check_valid(self):
         if not self.signature or len(self.signature) == 0:
             print("No signature is this transaction!")
             return False
 
         h = SHA256.new()
-        h.update("self.id + self.fromAddress + self.toAddress + str(self.amount)".encode())
-        verify_sig(h, self.signature, node_id)  # Throws error if signature is invalid
+        h.update(self.transaction_content().encode())
+
+        try:
+            verify_sig(h, self.signature, self.node_id)  # Throws error if signature is invalid
+        except Exception:
+            return False
+
         return True
 
+    def transaction_content(self):
+        return "{}{}{}{}{}".format(self.id, self.node_id, self.fromAddress, self.toAddress, self.amount)
+
 
 class Block:
 
-    def __init__(self, timestamp, transactions, index):
+    def __init__(self, timestamp, transactions, index, previous_hash=""):
+        self.check_arguments(timestamp, transactions, index, previous_hash)
         self.timestamp = timestamp
         self.transactions = transactions
         self.index = index
-        self.previousHash = ""
+        self.previousHash = previous_hash
         self.currentHash = ""
         self.nonce = 0
 
     def __repr__(self):
         return self.timestamp + self.transactions + "Previous hash: " + self.previousHash + self.currentHash
 
+    def check_arguments(self, timestamp, transactions, index, previous_hash):
+        if not timestamp or not transactions:
+            raise Exception("Block must have a timestamp, one or more transactions and the previous block hash")
+
+        # TODO - Timestamp check fails when using datetime.now, find the correct type
+        # if not isinstance(timestamp, type(datetime.now)):
+        #    raise Exception("Timestamp must be a datetime")
+
+        if index < 0 or not type(index) is int:
+            raise Exception("Index must be a positive integer greater or equal to 0")
+
+        try:
+            self.has_valid_transactions(transactions)
+        except:
+            raise Exception("Invalid Transactions")
+
+        # TODO - Add check for previous hash to be empty only for first block
```

Comment:
Those todos are excellent catches: perhaps add one issue if you are not planning to address them atm

Reply:
Issues added!